### ISSUE TITLE
FsEvents daemon crashing / not working

### ISSUE THREAD

**Issue by michaellopez** [codekitchen/dinghy#181]
I'm seeing this error:
```sh
$ cat /usr/local/Cellar/dinghy/4.3.2/local/var/dinghy-FsEvents.output
/Library/Ruby/Gems/2.0.0/gems/net-ssh-2.9.2/lib/net/ssh/transport/session.rb:188:in `block in poll_message': disconnected: Too many authentication failures (2) (Net::SSH::Disconnect)
	from /Library/Ruby/Gems/2.0.0/gems/net-ssh-2.9.2/lib/net/ssh/transport/session.rb:178:in `loop'
	from /Library/Ruby/Gems/2.0.0/gems/net-ssh-2.9.2/lib/net/ssh/transport/session.rb:178:in `poll_message'
	from /Library/Ruby/Gems/2.0.0/gems/net-ssh-2.9.2/lib/net/ssh/transport/session.rb:163:in `next_message'
	from /Library/Ruby/Gems/2.0.0/gems/net-ssh-2.9.2/lib/net/ssh/authentication/session.rb:95:in `block in next_message'
	from /Library/Ruby/Gems/2.0.0/gems/net-ssh-2.9.2/lib/net/ssh/authentication/session.rb:94:in `loop'
	from /Library/Ruby/Gems/2.0.0/gems/net-ssh-2.9.2/lib/net/ssh/authentication/session.rb:94:in `next_message'
	from /Library/Ruby/Gems/2.0.0/gems/net-ssh-2.9.2/lib/net/ssh/authentication/methods/publickey.rb:53:in `authenticate_with'
	from /Library/Ruby/Gems/2.0.0/gems/net-ssh-2.9.2/lib/net/ssh/authentication/methods/publickey.rb:20:in `block in authenticate'
	from /Library/Ruby/Gems/2.0.0/gems/net-ssh-2.9.2/lib/net/ssh/authentication/key_manager.rb:112:in `block in each_identity'
	from /Library/Ruby/Gems/2.0.0/gems/net-ssh-2.9.2/lib/net/ssh/authentication/key_manager.rb:104:in `each'
	from /Library/Ruby/Gems/2.0.0/gems/net-ssh-2.9.2/lib/net/ssh/authentication/key_manager.rb:104:in `each_identity'
	from /Library/Ruby/Gems/2.0.0/gems/net-ssh-2.9.2/lib/net/ssh/authentication/methods/publickey.rb:19:in `authenticate'
	from /Library/Ruby/Gems/2.0.0/gems/net-ssh-2.9.2/lib/net/ssh/authentication/session.rb:79:in `block in authenticate'
	from /Library/Ruby/Gems/2.0.0/gems/net-ssh-2.9.2/lib/net/ssh/authentication/session.rb:66:in `each'
	from /Library/Ruby/Gems/2.0.0/gems/net-ssh-2.9.2/lib/net/ssh/authentication/session.rb:66:in `authenticate'
	from /Library/Ruby/Gems/2.0.0/gems/net-ssh-2.9.2/lib/net/ssh.rb:211:in `start'
	from /Library/Ruby/Gems/2.0.0/gems/fsevents_to_vm-1.1.1/lib/fsevents_to_vm/ssh_emit.rb:24:in `ssh'
	from /Library/Ruby/Gems/2.0.0/gems/fsevents_to_vm-1.1.1/lib/fsevents_to_vm/ssh_emit.rb:14:in `event'
	from /Library/Ruby/Gems/2.0.0/gems/fsevents_to_vm-1.1.1/lib/fsevents_to_vm/cli.rb:30:in `block in start'
	from /Library/Ruby/Gems/2.0.0/gems/fsevents_to_vm-1.1.1/lib/fsevents_to_vm/watch.rb:14:in `block (2 levels) in run'
	from /Library/Ruby/Gems/2.0.0/gems/fsevents_to_vm-1.1.1/lib/fsevents_to_vm/watch.rb:12:in `each'
	from /Library/Ruby/Gems/2.0.0/gems/fsevents_to_vm-1.1.1/lib/fsevents_to_vm/watch.rb:12:in `block in run'
	from /Library/Ruby/Gems/2.0.0/gems/rb-fsevent-0.9.6/lib/rb-fsevent/fsevent.rb:45:in `call'
	from /Library/Ruby/Gems/2.0.0/gems/rb-fsevent-0.9.6/lib/rb-fsevent/fsevent.rb:45:in `run'
	from /Library/Ruby/Gems/2.0.0/gems/fsevents_to_vm-1.1.1/lib/fsevents_to_vm/watch.rb:17:in `run'
	from /Library/Ruby/Gems/2.0.0/gems/fsevents_to_vm-1.1.1/lib/fsevents_to_vm/cli.rb:24:in `start'
	from /Library/Ruby/Gems/2.0.0/gems/thor-0.19.1/lib/thor/command.rb:27:in `run'
	from /Library/Ruby/Gems/2.0.0/gems/thor-0.19.1/lib/thor/invocation.rb:126:in `invoke_command'
	from /Library/Ruby/Gems/2.0.0/gems/thor-0.19.1/lib/thor.rb:359:in `dispatch'
	from /Library/Ruby/Gems/2.0.0/gems/thor-0.19.1/lib/thor/base.rb:440:in `start'
	from /Library/Ruby/Gems/2.0.0/gems/fsevents_to_vm-1.1.1/exe/fsevents_to_vm:4:in `<top (required)>'
	from /usr/local/bin/fsevents_to_vm:23:in `load'
	from /usr/local/bin/fsevents_to_vm:23:in `<main>'
```

Trying a restart works:
```sh
$ dinghy restart
Stopping the FsEvents daemon
Stopping the dinghy VM...
Stopping "dinghy"...
Machine "dinghy" was stopped.
Stopping NFS daemon, this will require sudo
Stopping the DNS daemon
Starting the dinghy VM...
Starting NFS daemon, this will require sudo
Waiting for NFS daemon...
Mounting NFS /Users/michael
Starting the FsEvents daemon
Starting the DNS daemon
Starting the HTTP proxy
c87038997bbb0933ddf40837b78f930284cb1adb05ecff06c8b2b13a166c1930
  VM: running
 NFS: running
FSEV: running
 DNS: running
HTTP: running

Your environment variables are already set correctly.
```
Status shows OK
```sh
$ dinghy status
  VM: running
 NFS: running
FSEV: running
 DNS: running
HTTP: running
```

But as soon as I edit a file (no container running except dinghy proxy/dns container) I get this:
```sh
$ dinghy status
  VM: running
 NFS: running
FSEV: stopped
 DNS: running
HTTP: running

FsEvents failed to run
details available in log file: /usr/local/Cellar/dinghy/4.3.2/local/var/dinghy-FsEvents.output

Your environment variables are already set correctly.
```

And the contents of `/usr/local/Cellar/dinghy/4.3.2/local/var/dinghy-FsEvents.output` is at the top of this issue. Full circle :)

I haven't relied on FsEvents before, only NFS so I can't say for sure its work before for my setup. But I did install the beta for Docker for Mac and I ended up having to destroy my dinghy VM and create it again (I don't exactly know why, but somehow network got b0rked in the VM I think). Maybe this can be the cause? As it seems to be unable to auth via public_key over SSH. I'm guessing dinghy is trying to ssh into the VM (which now isn't the same VM as when I installed and later upgraded dinghy with)

Let me know if I can help more.

**Comment by michaellopez** [comment]
Oh, `dinghy ssh` works without problem. Might be good to know.

**Comment by michaellopez** [comment]
Having debugged a bit more and the variables look correct in `lib/fsevents_to_vm/ssh_emit.rb`

```
ip: 192.168.99.100
username: docker
identity file: /Users/michael/.docker/machine/machines/dinghy/id_rsa
```
Normal SSH works:
```sh
$ ssh -T -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -i /Users/michael/.docker/machine/machines/dinghy/id_rsa docker@192.168.99.100
Warning: Permanently added '192.168.99.100' (RSA) to the list of known hosts.
Boot2Docker version 1.11.1, build HEAD : 7954f54 - Wed Apr 27 16:36:45 UTC 2016
Docker version 1.11.1, build 5604cbe
^CKilled by signal 2.
```
I thought that maybe the host signature change was causing the problem, but saw that `:paranoid` was set to `false` for `Net::SSH`. Will dig further...

**Comment by michaellopez** [comment]
More searching led me to this remark: http://makandracards.com/makandra/512-how-to-fix-too-many-authentic-authentication-failures-with-ssh-and-or-capistrano#painful-fix-for-capistrano-only-monkey-patch-your-deployrb

Will give it a test...

I have several keys (1 for each service) and my ssh config is heavily using `IdentitiesOnly yes` and specifying which key to use for each service. So this might actually be it...

**Comment by michaellopez** [comment]
Yes, that was the cause for me...PR incoming...

**Comment by michaellopez** [comment]
Tried with my PR over at https://github.com/codekitchen/fsevents_to_vm/pull/6 and fsevents daemon now stays up and running and [nodemon](https://github.com/remy/nodemon) now triggers perfectly inside container without polling.

**Comment by codekitchen** [comment]
Thanks again, this is on master and I'll cut a release soon

**Comment by michaellopez** [comment]
Sounds good!

**Comment by mickaelperrin** [comment]
Hi,

Dinghy used to work like a charm, but since this evening I can't get it to work.

DNS resolution fails and FSevent crashes with the same error: `/Library/Ruby/Gems/2.0.0/gems/net-ssh-2.9.4/lib/net/ssh/transport/session.rb:188:in block in poll_message': disconnected: Too many authentication failures (2) (Net::SSH::Disconnect)`.

I use the 4.4.0 version.

I am currently experiencing DSL issues and can't rebuild the dinghy vm.

Any hints ?

**Comment by michaellopez** [comment]
@mickaelperrin What is the output of `sudo GEM_PATH=/Library/Ruby/Gems/2.0.0 GEM_HOME=/Library/Ruby/Gems/2.0.0 gem list`?

**Comment by michaellopez** [comment]
@mickaelperrin Also note that failing DNS resolution is not something that is caused by fsevents. This is only for forwarding file system events to the container.

**Comment by mickaelperrin** [comment]
Hi, thanks for your support.

This morning dinghy proxy runs normally but fevents still crashes. It was working very well (as usual) until yesterday when I left my office. At home, I woke up my computer and it was not working any more... crazy...

Here is the information you asked:

```
*** LOCAL GEMS ***

bigdecimal (1.2.6)
bundler (1.11.2)
fsevents_to_vm (1.1.2, 1.1.1)
io-console (0.4.3)
json (1.8.1)
net-ssh (2.9.4)
psych (2.0.8)
rake (10.4.2)
rb-appscript (0.6.1)
rb-fsevent (0.9.7)
rdoc (4.2.0)
rubygems-update (2.6.2)
thor (0.19.1)
require 'dinghy/daemon'

class FseventsToVm
  include Dinghy::Daemon
  INSTALL_PATH = Dinghy.brew+"bin"
  BIN_PATH = INSTALL_PATH+"fsevents_to_vm"
  VERSION = "~> 1.1.2"

  attr_reader :machine

  def initialize(machine)
    @machine = machine
  end

  def up
    install_if_necessary!
    increase_inotify_limit
    super
  end

  def plist_name
    "dinghy.fsevents_to_vm.plist"
  end

  def name
    "FsEvents"
  end

  protected

  def install_if_necessary!
    %x{/System/Library/Frameworks/Ruby.framework/Versions/Current/usr/bin/gem list -i -v '#{VERSION}' fsevents_to_vm}
    return if $?.success? and File.exists? BIN_PATH
    puts "Installing fsevents_to_vm, this will require sudo"
    system!("installing", "sudo", "/System/Library/Frameworks/Ruby.framework/Versions/Current/usr/bin/gem", "install", "--no-rdoc", "--no-ri", "-n", INSTALL_PATH, "fsevents_to_vm", "-v", VERSION)
  end

  def increase_inotify_limit
    machine.ssh("echo fs.inotify.max_user_watches=524288 | sudo tee -a /etc/sysctl.conf > /dev/null")
    machine.ssh("sudo sysctl -p > /dev/null")
  end

  def command
    %W[
      #{BIN_PATH}
      start
      --ssh-identity-file=#{machine.ssh_identity_file_path}
      --ssh-ip=#{machine.vm_ip}
    ]
  end
end
```
Note that I don't use `4.4.0` version but `HEAD`.

I will try to update dinghy and rebuild the all vm this morning.

**Comment by mickaelperrin** [comment]
Just did a new install with `brew reinstall --HEAD dinghy` and a `dinghy destroy` followed by `dinghy create -m 2048 -c 2 -d 40000 -p vmware`. The result of `dinghy status` is still the same `/Library/Ruby/Gems/2.0.0/gems/net-ssh-2.9.4/lib/net/ssh/transport/session.rb:188:in `block in poll_message': disconnected: Too many authentication failures (2) (Net::SSH::Disconnect)`

**Comment by mickaelperrin** [comment]
Mmm... that SSH failure makes me think of a bug I had on NodeJS related with SSH also. As soon as my SSH agent has more than a few keys, ssh authentication was failing because it tries sequentially each key in the agent and fails before using the good one.

I purged my agent with `ssh-add -D`, restart dinghy and fsEvent doesn't crash at start anymore.

I don't know if some additional configuration can prevent that behavior... I tried to play previously with `IdentitiesOnly` but with no success on that problem.

**Comment by mickaelperrin** [comment]
In fact, my version of fsevents_to_vm didn't have the patch https://github.com/codekitchen/fsevents_to_vm/pull/6 applied. Don't know why...

When I do a `brew reinstall --HEAD dinghy` does it update also the related projects to the good versions accordingly ? Is it related to the fact that the version number of fsevents_to_vm wasn't updated ?

**Comment by codekitchen** [comment]
That fix is in fsevents_to_vm version 1.1.2, which dinghy started using in verison 4.4.0. A `reinstall --HEAD` _should_ pick that up, an install of the latest Dinghy release 4.4.0 would work too. What does `dinghy version` say?

**Comment by mickaelperrin** [comment]
Dinghy version says `4.4.0`. I can confirm that fsevents_to_vm 1.1.2 is installed. Don't understand why it didn't work previously, but that's good now.
